spec/services/destroy_task_spec.rb

```diff
@@ -76,18 +76,36 @@
     let!(:p1) { Fabricate(:post, topic: t1) }
     let!(:p2) { Fabricate(:post, topic: t1) }
     let!(:p3) { Fabricate(:post, topic: t2) }
+    let!(:p4) { Fabricate(:post, topic: t2) }
 
     before { p2.trash! }
 
-    it "destroys posts listed and creates staff action logs" do
-      expect { task.destroy_posts([p2.id, p3.id], require_confirmation: false) }.to change {
-        Post.with_deleted.count
-      }.by(-2).and change { UserHistory.pluck(:action) }.from([]).to(
-              [
-                UserHistory.actions[:delete_post_permanently],
-                UserHistory.actions[:delete_topic_permanently],
-              ],
-            )
+    context "when permanent deletion is enabled" do
+      it "destroys posts listed and creates staff action logs" do
+        SiteSetting.can_permanently_delete = true
+
+        expect {
+          task.destroy_posts([p2.id, p3.id, p4.id], require_confirmation: false)
+        }.to change { Post.with_deleted.count }.by(-3).and change {
+                UserHistory.pluck(:post_id, :topic_id, :action)
+              }.from([]).to(
+                contain_exactly(
+                  [p2.id, nil, UserHistory.actions[:delete_post_permanently]],
+                  [nil, t2.id, UserHistory.actions[:delete_topic_permanently]],
+                  [p4.id, nil, UserHistory.actions[:delete_post_permanently]],
```

Comment:
We'd previously get a duplicate here, due to us manually deleting `p4`, and it also being deleted as part of the deletion of `t2`, triggered by deleting its first post.